docs/dev/RewriterToVisitorTransformGuide.md

```diff
@@ -0,0 +1,202 @@
+|SelectProjectionValue	|Projection.ProjectValue	|rewriteSelectProjectionValue	|transformProjectionProjectValue	|	|
+|SelectProjectionList	|Projection.ProjectList	|rewriteSelectProjectionList	|transformProjectionProjectList	|	|
+|FromSourceLet	|FromSource.Scan and FromSource.Unpivot	|rewriteFromSourceLet	|transformFromSourceScan and transformFromSourceUnpivot	|PartiqlAst does not have the supertype w/ Scan and Unpivot	|
+|FromSourceExpr	|FromSource.Scan	|rewriteFromSourceExpr	|transformFromSourceScan	|	|
+|ScopeQualifier.LEXICAL	|ScopeQualifier.LocalsFirst	|	|	|ScopeQualifier.UNQUALIFIED is the same	|
+
+## Converting Metas
+
+The constructor for metas is slightly different between ast’s MetaContainer and IonElement’s MetaContainer. Here is an example conversion:
+
+```kotlin
+metaContainerOf(UniqueNameMeta("foo"))
+```
+
+becomes 
+
+```kotlin
+ metaContainerOf(UniqueNameMeta.TAG to UniqueNameMeta("foo"))
+```
+
+
+Generally, you do not want to keep ast’s MetaContainer and have to use the conversion function ( [`toPartiQlMetaContainer`](https://github.com/partiql/partiql-lang-kotlin/blob/36fe55b6457f0dc82c02da6975358c01dd13aa3b/lang/src/org/partiql/lang/ast/StatementToExprNode.kt#L38))
+
+To add a new meta to an IonElement MetaContainer, you can use the `withMeta(metaKey: String, metaValue: Any)`
+function.
+
+To add multiple metas, you will need to add two MetaContainers together:
+
+```kotlin
+metas = oldMetaContainer + metaContainerOf(newMeta)
+```
+
+
+
+## Inner Rewrite Query
+
+In the rewriter code, you may encounter [`innerRewriteSelect`](https://github.com/partiql/partiql-lang-kotlin/blob/36fe55b6457f0dc82c02da6975358c01dd13aa3b/lang/src/org/partiql/lang/ast/passes/AstRewriterBase.kt#L151). This function calls the rewriter function for each clause of the SFW query. The direct equivalent in PIG is `transformExprSelect`.  In most cases you should be able to just call `transformExprSelect` directly or the base class’s transform, `super.transformExprSelect`.
+
+However, currently there are some slight differences between the order of transforms in `innerRewriteSelect` and `transformExprSelect`. `innerRewriteSelect` follows the traversal order (SQL semantic order) that is
+
+FROM → (FROM LET) → (WHERE) → (GROUP BY) → (HAVING) → PROJECTION → (LIMIT)
+
+`transformExprSelect` transforms the clauses in the order they are written for a PartiQL query, that is
+
+PROJECTION → FROM → (FROM LET) → (WHERE) → (GROUP BY) → (HAVING) → (LIMIT)
+
+This slight difference can lead to some different behaviors when translating `innerRewriteSelect` (e.g. StaticTypeRewriter). So to completely solve this issue, you can have your VisitorTransform implement [`VisitorTransformBase`](https://github.com/partiql/partiql-lang-kotlin/blob/master/lang/src/org/partiql/lang/eval/visitors/VisitorTransformBase.kt) and call `transformExprSelectEvaluationOrder` rather than `transformExprSelect` to get the same behavior.
+
```

Comment:
You might want to add another note in here about how `innerRewriteSelect` and `transformExprSelectEvaluationOrder` can be used to avoid infinite recursion in the case of nested rewriter/transform instances.

Reply:
Good point. Added in the latest commit.